zhusuan/distributions/multivariate.py

```diff
@@ -29,6 +33,158 @@
 ]
 
 
+class MultivariateNormalTriL(Distribution):
+    """
+    The class of multivariate normal distribution, where covariance is
+    parameterized with the lower triangular matrix in Cholesky decomposition,
+
+        .. math :: L \\text{s.t.} LL^T = \Sigma.
+
+    See :class:`~zhusuan.distributions.base.Distribution` for details.
+
+    :param mean: An N-D `float` Tensor. mean[..., i_{N-1}, :] corresponds to a
+        single distribution.
+    :param cov_tril: An (N+1)-D `float` Tensor. Each slice represents the lower
+        triangular matrix in the Cholesky decomposition.
+    :param group_ndims: A 0-D `int32` Tensor representing the number of
+        dimensions in `batch_shape` (counted from the end) that are grouped
+        into a single event, so that their probabilities are calculated
+        together. Default is 0, which means a single value is an event.
+        See :class:`~zhusuan.distributions.base.Distribution` for more detailed
+        explanation.
+    :param is_reparameterized: A Bool. If True, gradients on samples from this
+        distribution are allowed to propagate into inputs, using the
+        reparametrization trick from (Kingma, 2013).
+    :param use_path_derivative: A bool. Whether when taking the gradients
+        of the log-probability to propagate them through the parameters
+        of the distribution (False meaning you do propagate them). This
+        is based on the paper "Sticking the Landing: Simple,
+        Lower-Variance Gradient Estimators for Variational Inference"
+    :param check_numerics: Bool. Whether to check numeric issues.
+    """
+
+    def __init__(self,
+                 mean,
+                 cov_tril,
+                 group_ndims=0,
+                 is_reparameterized=True,
+                 use_path_derivative=False,
+                 check_numerics=False,
+                 **kwargs):
+        self._check_numerics = check_numerics
+        self._mean = assert_rank_at_least_one(mean, 'MVNTriL.mean')
+        self._n_dim = get_shape_at(self._mean, -1)
+        self._cov_tril = assert_rank_at_least(cov_tril, 2, 'MVNTriL.cov_tril')
+
+        # Check shape; fail asap
+        if self._mean.get_shape() and self._cov_tril.get_shape():
+            expected_shape = get_shape_list(self._mean) + [self._n_dim]
+            actual_shape = get_shape_list(self._cov_tril)
+            err_msg = ['MVNTriL.cov_tril should have compatible shape with',
+                       ' mean. Expected', tf.convert_to_tensor(expected_shape),
```

Comment:
This may fail if there is `None` in `expected_shape`. Use [assert_is_compatible_with](https://www.tensorflow.org/api_docs/python/tf/TensorShape#assert_is_compatible_with) or `is_compatible_with` to do this here.

Reply:
Right. But `assert_is_compatible_with` doesn't do dynamic check, and [this test](https://github.com/meta-inf/zhusuan/blob/d6f4b286b79adae59387341a0281967fb7541e00/tests/distributions/test_multivariate.py#L30) will fail. I fixed it another way, maybe you could check on it.

Reply:
But the `else` branch does dynamic check? So I thought the `if` branch is only for static check.

Reply:
I don't see why it fails.

Reply:
No, I meant when the shape is like `[None, 5]`, then `get_shape` does not convert to False. 
The behavior you was thinking of can be implemented with `TensorShape.is_fully_defined`. Then fewer static checks are done, but the code is neater.

Reply:
I see another way of doing this. You could throw out this if/else, and just use `assert_is_compatible_with` to check static shapes first (it will work with static shapes not fully defined) and then add dynamic asserts. That will solve the problem.

Reply:
Right. But as I said it won't check conflicts like `[None,5]` vs `[None,6]` before execution. Just to make sure, is it desirable?

Reply:
It will.
```python
a = tf.placeholder(tf.float32, [None, 5])
b = tf.placeholder(tf.float32, [None, 6])
a.shape.assert_is_compatible_with(b.shape)
# ==> ValueError: Shapes (?, 5) and (?, 6) are incompatible
```

Reply:
Seems I got something wrong. Sorry about this.